sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java

```diff
@@ -616,6 +624,58 @@ public void testRiskyConfigurationWarnsProperly() {
     p.run();
   }
 
+  @Test
+  public void testRiskyConfigurationWarnsProperlyWithNumShardsNotSet() {
+    int numElements = 1000;
+
+    PCollection<Long> input =
+        p.apply(
+                mkKafkaReadTransform(numElements, numElements, new ValueAsTimestampFn(), true, 0)
+                    .withConsumerConfigUpdates(
+                        ImmutableMap.of(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, true))
+                    .withoutMetadata())
+            .apply(Values.create());
+
+    addCountingAsserts(input, numElements);
+
+    kafkaIOExpectedLogs.verifyWarn(
+        "This will redistribute the load across the same number of shards as the Kafka source.");
```

Comment:
You're correct, I have removed this comment, and updated it to 

"This will create a key per record, which is sub-optimal for most use cases." -> this is still an implementation detail, but is a beam implementation detail rather than from the underlying runner.

Reply:
What I mean about the implementation detail is that it is Dataflow-specific that we need fewer keys instead of doing something intelligent like bundling across keys (which is very much allowed by the model).